File: upstream/nuvio-engine/platform/apple/Sources/NuvioEngine/NuvioModels.swift
import Foundation

public enum NuvioEventType: UInt32, Sendable, CaseIterable {
    case torrentAdded = 1
    case torrentMetadataReady = 2
    case torrentError = 3
    case streamPrepared = 4
    case torrentRemoved = 5
    case streamStopped = 6
    case diskCacheReclaimed = 7
}

public struct NuvioEvent: Sendable, Equatable {
    public let type: NuvioEventType
    public let sequence: UInt64
    public let requestID: UInt64
    public let droppedEvents: UInt64
    public let torrentID: String?
    public let message: String?
    public let fileIndex: UInt32?
    public let fileSize: UInt64
    public let streamID: String?
    public let streamURL: URL?
}

public struct NuvioTorrentFile: Sendable, Equatable {
    public let index: UInt32
    public let offset: UInt64
    public let size: UInt64
    public let path: String
    public let pathWasTruncated: Bool
}

public struct NuvioStream: Sendable, Equatable {
    public let id: String
    public let url: URL
    public let torrentID: String
    public let fileIndex: UInt32
    public let fileSize: UInt64
}

public struct NuvioStreamStats: Sendable, Equatable {
    public let fileIndex: UInt32
    public let fileSize: UInt64
    public let contiguousReadyBytes: UInt64
    public let verifiedFileBytes: UInt64
    public let deliveredBytes: UInt64

    public var bufferProgress: Double { ratio(contiguousReadyBytes, fileSize) }
    public var fileProgress: Double { ratio(verifiedFileBytes, fileSize) }
}

public struct NuvioEngineStats: Sendable, Equatable {
    public let activeTorrents: UInt32
    public let activeStreams: UInt32
    public let activeHTTPRequests: UInt32
    public let connectedPeers: UInt32
    public let connectedSeeds: UInt32
    public let knownPeers: UInt32
    public let connectCandidates: UInt32
    public let interestedPeers: UInt32
    public let unchokedPeers: UInt32
    public let downloadingPeers: UInt32
    public let snubbedPeers: UInt32
    public let pendingBlockRequests: UInt32
    public let targetBlockRequests: UInt32
    public let timedOutBlockRequests: UInt32
    public let pendingPieceReads: UInt32
    public let downloadRateBytesPerSecond: UInt64
    public let uploadRateBytesPerSecond: UInt64
    public let totalPayloadDownloadBytes: UInt64
    public let totalPayloadUploadBytes: UInt64
    public let memoryCacheCapacityBytes: UInt64
    public let memoryCacheUsedBytes: UInt64
    public let memoryCacheHits: UInt64
    public let memoryCacheMisses: UInt64
    public let memoryCacheEvictions: UInt64
    public let memoryCacheEntries: UInt64
    public let warmTorrents: UInt32
    public let quiescedTorrents: UInt32
    public let diskCacheCapacityBytes: UInt64
    public let diskCacheUsedBytes: UInt64
    public let diskCacheProtectedBytes: UInt64
    public let diskCacheEvictions: UInt64
    public let diskCacheReclaimedBytes: UInt64
    public let diskCacheIsOverBudget: Bool

    public init(
        activeTorrents: UInt32 = 0,
        activeStreams: UInt32 = 0,
        activeHTTPRequests: UInt32 = 0,
        connectedPeers: UInt32 = 0,
        connectedSeeds: UInt32 = 0,
        knownPeers: UInt32 = 0,
        connectCandidates: UInt32 = 0,
        interestedPeers: UInt32 = 0,
        unchokedPeers: UInt32 = 0,
        downloadingPeers: UInt32 = 0,
        snubbedPeers: UInt32 = 0,
        pendingBlockRequests: UInt32 = 0,
        targetBlockRequests: UInt32 = 0,
        timedOutBlockRequests: UInt32 = 0,
        pendingPieceReads: UInt32 = 0,
        downloadRateBytesPerSecond: UInt64 = 0,
        uploadRateBytesPerSecond: UInt64 = 0,
        totalPayloadDownloadBytes: UInt64 = 0,
        totalPayloadUploadBytes: UInt64 = 0,
        memoryCacheCapacityBytes: UInt64 = 0,
        memoryCacheUsedBytes: UInt64 = 0,
        memoryCacheHits: UInt64 = 0,
        memoryCacheMisses: UInt64 = 0,
        memoryCacheEvictions: UInt64 = 0,
        memoryCacheEntries: UInt64 = 0,
        warmTorrents: UInt32 = 0,
        quiescedTorrents: UInt32 = 0,
        diskCacheCapacityBytes: UInt64 = 0,
        diskCacheUsedBytes: UInt64 = 0,
        diskCacheProtectedBytes: UInt64 = 0,
        diskCacheEvictions: UInt64 = 0,
        diskCacheReclaimedBytes: UInt64 = 0,
        diskCacheIsOverBudget: Bool = false
    ) {
        self.activeTorrents = activeTorrents
        self.activeStreams = activeStreams
        self.activeHTTPRequests = activeHTTPRequests
        self.connectedPeers = connectedPeers
        self.connectedSeeds = connectedSeeds
        self.knownPeers = knownPeers
        self.connectCandidates = connectCandidates
        self.interestedPeers = interestedPeers
        self.unchokedPeers = unchokedPeers
        self.downloadingPeers = downloadingPeers
        self.snubbedPeers = snubbedPeers
        self.pendingBlockRequests = pendingBlockRequests
        self.targetBlockRequests = targetBlockRequests
        self.timedOutBlockRequests = timedOutBlockRequests
        self.pendingPieceReads = pendingPieceReads
        self.downloadRateBytesPerSecond = downloadRateBytesPerSecond
        self.uploadRateBytesPerSecond = uploadRateBytesPerSecond
        self.totalPayloadDownloadBytes = totalPayloadDownloadBytes
        self.totalPayloadUploadBytes = totalPayloadUploadBytes
        self.memoryCacheCapacityBytes = memoryCacheCapacityBytes
        self.memoryCacheUsedBytes = memoryCacheUsedBytes
        self.memoryCacheHits = memoryCacheHits
        self.memoryCacheMisses = memoryCacheMisses
        self.memoryCacheEvictions = memoryCacheEvictions
        self.memoryCacheEntries = memoryCacheEntries
        self.warmTorrents = warmTorrents
        self.quiescedTorrents = quiescedTorrents
        self.diskCacheCapacityBytes = diskCacheCapacityBytes
        self.diskCacheUsedBytes = diskCacheUsedBytes
        self.diskCacheProtectedBytes = diskCacheProtectedBytes
        self.diskCacheEvictions = diskCacheEvictions
        self.diskCacheReclaimedBytes = diskCacheReclaimedBytes
        self.diskCacheIsOverBudget = diskCacheIsOverBudget
    }
}

public struct NuvioEngineError: Error, LocalizedError, Sendable, Equatable {
    public let status: UInt32?
    public let message: String

    public var errorDescription: String? { message }

    public init(status: UInt32? = nil, message: String) {
        self.status = status
        self.message = message
    }
}

private func ratio(_ value: UInt64, _ total: UInt64) -> Double {
    guard total > 0 else { return 0 }
    return min(max(Double(value) / Double(total), 0), 1)
}

File: upstream/nuvio-engine/platform/android/engine/src/main/kotlin/com/nuvio/engine/NuvioModels.kt
package com.nuvio.engine

public enum class NuvioEventType(public val nativeValue: Int) {
    TorrentAdded(1),
    TorrentMetadataReady(2),
    TorrentError(3),
    StreamPrepared(4),
    TorrentRemoved(5),
    StreamStopped(6),
    DiskCacheReclaimed(7);

    internal companion object {
        fun fromNative(value: Int): NuvioEventType =
            entries.firstOrNull { it.nativeValue == value }
                ?: throw NuvioEngineException(-1, "unknown native event type $value")
    }
}

public data class NuvioEvent(
    val type: NuvioEventType,
    val sequence: Long,
    val requestId: Long,
    val droppedEvents: Long,
    val torrentId: String?,
    val message: String?,
    val fileIndex: Int?,
    val fileSize: Long,
    val streamId: String?,
    val streamUrl: String?,
)

public data class NuvioTorrentFile(
    val index: Int,
    val offset: Long,
    val size: Long,
    val path: String,
    val pathTruncated: Boolean,
)

public data class NuvioStream(
    val id: String,
    val url: String,
    val torrentId: String,
    val fileIndex: Int,
    val fileSize: Long,
)

public data class NuvioStreamStats(
    val fileIndex: Int,
    val fileSize: Long,
    val contiguousReadyBytes: Long,
    val verifiedFileBytes: Long,
    val deliveredBytes: Long,
    val activeDemands: Int = 0,
    val scheduledPieces: Int = 0,
    val blockingPieces: Int = 0,
    val primaryBlockingPiece: Int = -1,
    val secondaryBlockingPiece: Int = -1,
    val lastReadyPiece: Int = -1,
    val primaryDemandStart: Long = 0,
    val primaryDemandEnd: Long = 0,
    val secondaryDemandStart: Long = 0,
    val secondaryDemandEnd: Long = 0,
    val scheduleRevision: Long = 0,
) {
    public val bufferProgress: Float
        get() = ratio(contiguousReadyBytes, fileSize)

    public val fileProgress: Float
        get() = ratio(verifiedFileBytes, fileSize)
}

public data class NuvioEngineStats(
    val activeTorrents: Int = 0,
    val activeStreams: Int = 0,
    val activeHttpRequests: Int = 0,
    val connectedPeers: Int = 0,
    val connectedSeeds: Int = 0,
    val knownPeers: Int = 0,
    val connectCandidates: Int = 0,
    val interestedPeers: Int = 0,
    val unchokedPeers: Int = 0,
    val downloadingPeers: Int = 0,
    val snubbedPeers: Int = 0,
    val pendingBlockRequests: Int = 0,
    val targetBlockRequests: Int = 0,
    val timedOutBlockRequests: Int = 0,
    val connectingPeers: Int = 0,
    val handshakingPeers: Int = 0,
    val targetPiecePeers: Int = 0,
    val targetPieceUnchokedPeers: Int = 0,
    val targetPieceDownloadingPeers: Int = 0,
    val offTargetDownloadingPeers: Int = 0,
    val trackerReplyEvents: Int = 0,
    val trackerErrorEvents: Int = 0,
    val dhtReplyEvents: Int = 0,
    val trackerPeersReturned: Long = 0,
    val dhtPeersReturned: Long = 0,
    val peerConnectEvents: Long = 0,
    val peerDisconnectEvents: Long = 0,
    val peerDisconnectTimeouts: Long = 0,
    val peerDisconnectConnectFailures: Long = 0,
    val peerDisconnectRedundant: Long = 0,
    val peerDisconnectTurnover: Long = 0,
    val peerDisconnectOther: Long = 0,
    val torrentFinishedEvents: Long = 0,
    val pendingPieceReads: Int = 0,
    val downloadRateBytesPerSecond: Long = 0,
    val uploadRateBytesPerSecond: Long = 0,
    val totalPayloadDownloadBytes: Long = 0,
    val totalPayloadUploadBytes: Long = 0,
    val memoryCacheCapacityBytes: Long = 0,
    val memoryCacheUsedBytes: Long = 0,
    val memoryCacheHits: Long = 0,
    val memoryCacheMisses: Long = 0,
    val memoryCacheEvictions: Long = 0,
    val memoryCacheEntries: Long = 0,
    val warmTorrents: Int = 0,
    val quiescedTorrents: Int = 0,
    val diskCacheCapacityBytes: Long = 0,
    val diskCacheUsedBytes: Long = 0,
    val diskCacheProtectedBytes: Long = 0,
    val diskCacheEvictions: Long = 0,
    val diskCacheReclaimedBytes: Long = 0,
    val diskCacheOverBudget: Boolean = false,
)

public class NuvioEngineException(
    public val status: Int,
    message: String,
) : Exception(message)

private fun ratio(value: Long, total: Long): Float =
    if (total <= 0L) 0f else (value.toDouble() / total.toDouble()).coerceIn(0.0, 1.0).toFloat()
